File: Programms/R/FeedForward.R
library(e1071) # contains sigmoid. Please don't ask why it's called that, I don't know

FeedForward <- function(pInputSize = 784, pOutputSize = 10, numOfHiddenLayers = 1,
                       hiddenLayerSize = 100, activation = sigmoid, learningfactor = 0.01){
  envi <- environment()
  
  # data fields (private)
  inputSize <- pInputSize
  outputSize <- pOutputSize
  hiddenLayerWeights <- array(data = runif(hiddenLayerSize^2 * (numOfHiddenLayers - 1), min = -1),
                        dim = c(numOfHiddenLayers -1, hiddenLayerSize, hiddenLayerSize))
  intoHiddenWeights <- array(data = runif(inputSize * hiddenLayerSize, min = -1), dim = c(hiddenLayerSize, inputSize))
  outOfHiddenWeights <- array(data = runif(hiddenLayerSize * outputSize, min = -1), dim = c(outputSize, hiddenLayerSize))
  
  # functions (the equals sign is mandatory)
  me <- list(
    envi = envi,
    
    getEnvironment = function(){
      return(get("envi",envi))
    },
    
    getInputSize = function(){
      return(get("inputSize",envi))
    },
    
    getOutputSize = function(){
      return(get("outputSize",envi))
    },
    
    evaluate = function(input) {
      if(!is.array(input)) stop("input must be an array")
      if(!identical( dim(input) , as.integer(c(inputSize)))) 
        stop(cat("Dimension mismatch in evaluate! Dimension of input needs to be", inputSize, "x 1"))
      output <- intoHiddenWeights %*% input
      output <- activation(output)
      if(numOfHiddenLayers > 1){
        for(layer in 1:numOfHiddenLayers) {
          output <- hiddenLayerWeights[layer,,] %*% output
          output <- activation(output)
        }
      }
      output <- outOfHiddenWeights %*% output
      return(output)
    },
    
    classify = function(input) {
      if(!is.array(input)) stop("input must be an array")
      if(!identical( dim(input) , as.integer(c(inputSize)))) 
        stop(cat("Dimension mismatch in classify! Dimension of input needs to be", inputSize, "x 1"))
      result <- me$evaluate(input)
      return(which.max(result))
    },
    
    train = function(input, output) {
      if(!is.array(inputData)) stop("inputData must be an array")
      if(!is.array(outputData)) stop("outputData must be an array")
      if(!identical( dim(inputData) , as.integer(c(inputSize)))) 
        stop(cat("Dimension mismatch in classify! Dimension of inputData needs to be", inputSize, "x 1"))
      if(!identical( dim(outputData) , as.integer(c(outputSize)))) 
        stop(cat("Dimension mismatch in classify! Dimension of inputData needs to be", inputSize, "x 1"))
      result <- me$evaluate(inputData)
      
      # Deltaregel
      d_output <- array(dim = outputSize)
      d <- array(dim = c(numOfHiddenLayers, hiddenLayerSize))
      output_gradient <- array(dim = c(outputSize, hiddenLayerSize))
      input_gradient <- array(dim = c(hiddenLayerSize, inputSize))
      hidden_gradient <- array(dim = c(numOfHiddenLayers -1, hiddenLayerSize, hiddenLayerSize))
      for(outputNeuron in 1:nrow(gradient)){
        for(inputNeuron in 1:ncol(gradient)){
          # \Delta w_ik = e * d_i * a_k
          # e is out of loop
          d <- output[outputNeuron] - result[outputNeuron]
          gradient[outputNeuron, inputNeuron] <- d * input[inputNeuron]
        }
      }
      weights <- weights + learnfactor * gradient
    }
    
  )
  
  assign('this', me, envir = envi)
  class(me) <- append(class(me), 'Perceptron')
  return(me)
}

File: Programms/Python/ownNets/FeedForward.py
import numpy as np

from ownNets.NN import MyNeuronalNet

import math


def sigmoid(x):
    return 1 / (1 + math.exp(-x))


def d_sigmoid(x):
    return sigmoid(x) * (1 - sigmoid(x))


class FeedForward(MyNeuronalNet):
    def __init__(self, input_size=784, output_size=10, num_of_hidden_layers=1,
                 hidden_layer_size=100, activation=sigmoid, activation_derivative=d_sigmoid, learning_speed=0.01):
        self.input_size = input_size
        self.output_size = output_size
        self.hidden_layer_activation = np.zeros((num_of_hidden_layers, hidden_layer_size))
        self.hidden_layer_weights = np.random.rand(num_of_hidden_layers - 1, hidden_layer_size,
                                                   hidden_layer_size) * 2 - 1
        self.into_hidden_weights = np.random.rand(hidden_layer_size, input_size) * 2 - 1
        self.out_of_hidden_weights = np.random.rand(output_size, hidden_layer_size) * 2 - 1
        self.activation = activation
        self.activation_derivative = activation_derivative
        self.learning_speed = learning_speed
        self.hidden_layer_size = hidden_layer_size
        self.num_of_hidden_layers = num_of_hidden_layers

    def evaluate(self, input_data):
        self.hidden_layer_activation[0] = self.activation(self.into_hidden_weights @ input_data)
        for i in range(1, self.num_of_hidden_layers - 1):
            self.hidden_layer_activation[i] = \
                self.activation(self.hidden_layer_weights[i] @ self.hidden_layer_activation[i - 1])
        output = self.activation(self.out_of_hidden_weights @ self.hidden_layer_activation[1])
        self.hidden_layer_activation = self.activation_derivative(self.hidden_layer_activation)
        return output

    def train(self, input_data, output_data):
        result = self.evaluate(input_data)
        # Backpropagation
        delta = np.zeros((self.num_of_hidden_layers, self.hidden_layer_size))

        # find delta for output neurons
        delta_output = self.activation_derivative(output_data - result)

        # find d for the last hidden layer
        for hidden_neuron in range(self.hidden_layer_size):
            delta[self.num_of_hidden_layers-2, hidden_neuron] = self.out_of_hidden_weights @ delta_output # to do
        pass
